### PR TITLE
Add file encryption events on file upload.

### PR DESCRIPTION
**Pull request by MartynasTreinys**
Add file decryption on file demand.
Add command to encrypt existing files.

### CODE REVIEW

File: src/AppBundle/Lasf/BinaryFileProvider.php [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    ) {
+        $this->em = $em;
+        $this->authorizationChecker = $authorizationChecker;
+        $this->crypt = $crypt;
+        $this->user = $tokenStorage->getToken()->getUser();
+        $this->webRoot = $webRoot;
+    }
+
+    /**
+     * Fetches
```
Comment (vytautas-k) [inline src/AppBundle/Lasf/BinaryFileProvider.php]: Comments or method descriptions should be meaningful. One word doesn't provide much information.

File: src/AppBundle/Lasf/BinaryFileProvider.php [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        EntityManagerInterface $em,
+        AuthorizationCheckerInterface $authorizationChecker,
+        CryptInterface $crypt,
+        TokenStorageInterface $tokenStorage,
+        $webRoot
+    ) {
+        $this->em = $em;
+        $this->authorizationChecker = $authorizationChecker;
+        $this->crypt = $crypt;
+        $this->user = $tokenStorage->getToken()->getUser();
```
Comment (vytautas-k) [inline src/AppBundle/Lasf/BinaryFileProvider.php]: I believe there can be cases when `getToken()` won't return user object (if class is initiated via command) or will return string (if user is not authenticated).

File: src/AppBundle/Lasf/BinaryFileProvider.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+     * @param string $path A path relative to 'storage_root' directory
+     * @return array
+     */
+    public function find($path)
+    {
+        $this->denyAccessUnlessGranted($path);
+
+        $fileContents = file_get_contents($this->webRoot . DIRECTORY_SEPARATOR . $path);
+
+        if (!($binary = $this->crypt->decrypt($fileContents))) {
```
Comment (vytautas-k) [inline src/AppBundle/Lasf/BinaryFileProvider.php]: I would recommend against assigning variables directly in if clauses as it's more difficult to read code and do not miss it.

File: src/AppBundle/Twig/FileExtension.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        return 'file';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFunctions()
+    {
+        return [
+            new \Twig_SimpleFunction('file', [$this, 'getFileUrl'])
```
Comment (vytautas-k) [inline src/AppBundle/Twig/FileExtension.php]: Comma should be added at the end of line in array to make GIT track only newly added elements in array.

File: src/AppBundle/Twig/FileExtension.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * @param FileUpload $fileUpload
+     * @param string $fieldName
+     * @return string
+     */
+    public function getFileUrl(FileUpload $fileUpload, $fieldName = 'file')
+    {
+        return $this->router->generate('app_fileuploads_download', [
+            'path' => $this->helper->asset($fileUpload, $fieldName)
```
Comment (vytautas-k) [inline src/AppBundle/Twig/FileExtension.php]: Same with comma in array here

File: src/AppBundle/Resources/config/listener.yml [hunk 6]
```diff
@@ -70,4 +70,11 @@ services:
...
-      - { name: doctrine.event_listener, event: onFlush }
\ No newline at end of file
+      - { name: doctrine.event_listener, event: onFlush }
+
+  app.file_upload_vich.listener:
+    class: AppBundle\EventListener\FileUploadVichListener
+    arguments:
+      - @app.crypt
+    tags:
+      - { name: kernel.event_listener, event: vich_uploader.pre_upload }
```
[the remark this answers is not in the document]
Comment (MartynasTreinys) [reply src/AppBundle/Resources/config/listener.yml]: It doesn't listen to events if I change it's name.

**Pull request merged**